models/articlesModels.js

```diff
@@ -55,26 +59,55 @@ exports.updateArticleById = ({ article_id }, { inc_votes }) => {
     });
 };
 
-exports.fetchArticles = () => {
+exports.fetchArticles = async (query) => {
+  let isInvalid = false;
+  const validQueries = ["sort_by", "order", "topic"];
+  Object.keys(query).forEach((query) => {
+    if (!validQueries.includes(query)) {
+      isInvalid = true;
+    }
+  });
+  if (isInvalid) {
+    return Promise.reject({ status: 400, msg: "invalid query" });
+  }
+
+  const validSortBy = [
+    "author",
+    "title",
+    "article_id",
+    "topic",
+    "created_at",
+    "votes",
+    "comment_count",
+  ];
+  const validOrder = ["asc", "desc"];
+
+  const { sort_by = "created_at", order = "desc", topic } = query;
+
+  let queryStr = `SELECT articles.article_id, articles.author, articles.title, articles.topic, articles.created_at, articles.votes, CAST (COUNT(comments.comment_id) AS INT) AS comment_count FROM articles LEFT JOIN comments ON articles.article_id = comments.article_id `;
+
+  if (topic !== undefined) {
+    await checkExists("topics", "slug", topic).catch(() => {
+      return Promise.reject({ status: 400, msg: "topic does not exist" });
+    });
+    queryStr += ` WHERE articles.topic LIKE '${topic}' `;
+  }
+
+  if (!validSortBy.includes(sort_by.toLowerCase())) {
+    return Promise.reject({ status: 400, msg: "invalid sort_by" });
+  }
+  if (!validOrder.includes(order.toLowerCase())) {
+    return Promise.reject({ status: 400, msg: "invalid order" });
+  }
+  queryStr += `GROUP BY articles.article_id ORDER BY ${sort_by} ${order};`;
+
   return db
-    .query(
-      `
-        SELECT articles.article_id, articles.author, 
-        articles.title,
-        articles.topic, 
-        articles.created_at, articles.votes,
-        CAST (COUNT(comments.comment_id) AS INT)
-        AS comment_count
-        FROM articles
-        LEFT JOIN comments
-        ON articles.article_id = comments.article_id
-        GROUP BY articles.article_id
-        ORDER BY articles.created_at DESC;`
-    )
-    .then((result) => {
-      return result.rows;
+    .query(queryStr)
+    .then(({ rows }) => {
+      return rows;
     })
     .catch((err) => {
```

Comment:
we should do our catch either in our controller or models and stick with one.

Reply:
So if `db.query` returns an error and I don't have a `catch` block in my models, will the `catch` in my controllers be able to catch it?